aws/resource_aws_acm_certificate_test.go

```diff
@@ -0,0 +1,311 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/acm"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"os"
+	"regexp"
+)
+
+func TestAccAwsAcmResource_certificateIssuingFlow(t *testing.T) {
+	var conf acm.DescribeCertificateOutput
+	var tags acm.ListTagsForCertificateOutput
+	var confAfterValidation acm.DescribeCertificateOutput
+	if os.Getenv("ACM_CERTIFICATE_ROOT_DOMAIN") == "" {
+		t.Skip("Environment variable ACM_CERTIFICATE_ROOT_DOMAIN is not set")
+	}
+
+	root_zone_domain := os.Getenv("ACM_CERTIFICATE_ROOT_DOMAIN")
+	domain := fmt.Sprintf("certtest.%s", root_zone_domain)
+	sanDomain := fmt.Sprintf("certtest2.%s", root_zone_domain)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAcmCertificateDestroy,
+		Steps: []resource.TestStep{
+			// Test that we can request a certificate
+			resource.TestStep{
+				Config: testAccAcmCertificateConfig(domain, sanDomain),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAcmCertificateExists("aws_acm_certificate.cert", &conf, &tags),
+					testAccCheckAcmCertificateAttributes("aws_acm_certificate.cert", &conf, domain, sanDomain, "PENDING_VALIDATION"),
+					testAccCheckTagsACM(&tags.Tags, "Hello", "World"),
+					testAccCheckTagsACM(&tags.Tags, "Foo", "Bar"),
+				),
+			},
+			// Test that we can change the tags
+			resource.TestStep{
+				Config: testAccAcmCertificateConfigWithChangedTags(domain, sanDomain),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAcmCertificateExists("aws_acm_certificate.cert", &conf, &tags),
+					testAccCheckAcmCertificateAttributes("aws_acm_certificate.cert", &conf, domain, sanDomain, "PENDING_VALIDATION"),
+					testAccCheckTagsACM(&tags.Tags, "Environment", "Test"),
+					testAccCheckTagsACM(&tags.Tags, "Foo", "Baz"),
+				),
+			},
+			// Test that validation times out if certificate can't be validated
```

Comment:
I agree in principle, there's a catch though: Certificate requests have a [limit](https://docs.aws.amazon.com/acm/latest/userguide/acm-limits.html) (twice the number of ACM certificates per region per year so 200 by default but maybe a lot lower) that you hit quickly if you run a lot of tests. It can be increased with approval from the service team but I'm not sure by how much (I got 400 requests per region now).

If we want to run these tests in CI, the HashiCorp account possibly needs a pretty high limit and every test that requests a certificate increases it.